graph.c

```diff
@@ -0,0 +1,350 @@
+
+                return;
+            }
+        }
+    }
+    else {
+        graph_add_neighbors_table(src);
+    }
+
+    graph_add_new_neighbor(src->neighbors, to_vertice, weight);
+
+}
+
+/*
+    erase edge from_vertice to_vertice if there is an edge
+    namely delete neighbor of from_vertice vertice
+*/
+bool graph_erase_edge(Graph *self, int from_vertice, int to_vertice) {
+    Graph_vertice* src;
+    Graph_vertice* dest;
+    int stub_weight = 1;
+
+    src = graph_get_vertice(self, from_vertice);
+    dest = graph_get_vertice(self, to_vertice);
+
+    graph_edge_function_check_pararameters(self, src, dest, from_vertice, to_vertice, stub_weight);
+
+
+    if (dest->neighbors != NULL) {
+        if (dest->neighbors->len > 0) {
+            Graph_vertice_neighbor* n;
+
+            if((n = graph_find_neighbor(dest->neighbors, from_vertice)) != NULL) {
+                n = NULL;
+
+                fprintf(stderr, "Path is found in to_vertice. Are index positions correct?");
+                abort();
+            }
+            n = NULL;
+        }
+    }
+
+    if (src->neighbors != NULL) {
+        if (src->neighbors->len > 0) {
+            Graph_vertice_neighbor* n;
+
+            if((n = graph_find_neighbor(src->neighbors, to_vertice)) != NULL) {
+                return g_ptr_array_remove(src->neighbors, n);
+            }
+            return FALSE;
+        }
+        return FALSE;
+    }
+    return FALSE;
+}
+
+static bool graph_is_null_pointer(Graph *self) {
+    return self == NULL;
+}
+
+/*
+    return pointer to vertice from graph_vertices_table if there is a vertice
+*/
+Graph_vertice* graph_get_vertice(Graph* self, int index) {
+    int* p_index;
+
+    if (graph_is_null_pointer(self)) {
+        fprintf(stderr, "graph pointer is equal to NULL");
+        abort();
+    }
+
+    if ((p_index = (int *) malloc(sizeof(index))) == NULL) {
+        fprintf(stderr, "memalloc for index failed");
+        abort();
+    }
+    *p_index = index;
+
+    Graph_vertice* p_vertice = g_hash_table_lookup(self->vertices_table, p_index);
+    free(p_index);
+
+    return p_vertice;
+}
+
+/*
+    erase vertice from neighbors table of other vertices, then
+    erase vertice from graph_vertices_table
+*/
+bool graph_erase_vertice(Graph* self, int index) {
+    int* p_index;
+
+    if (graph_is_null_pointer(self)) {
+        fprintf(stderr, "graph pointer is equal to NULL");
+        abort();
+    }
+
+    if ((p_index = (int *) malloc(sizeof(index))) == NULL) {
+        fprintf(stderr, "memalloc for index failed");
+        abort();
+    }
+    *p_index = index;
+
+    bool is_deleted = g_hash_table_remove(self->vertices_table, p_index);
+    free(p_index);
+
+    return is_deleted;
+}
+
+/*
+    add a new vertice without neighbors to graph_vertices_table
+*/
+void graph_add_vertice(Graph* self, int index) {
+    bool is_in_table;
+    int* p_index;
+    Graph_vertice *new_vertice;
+
+    if (graph_is_null_pointer(self)) {
```

Comment:
oh god, KISS

Reply:
DRY
Agreed here, too simple

Reply:
Although it catches my eye as duplication right

Reply:
It's fine but to check arguments for null. fopen and printf don't expect null as well